EC_finetune/senders.py

```diff
@@ -132,40 +141,75 @@ def forward(
-                image_hidden.device
-            )
-            for i in range(self.unroll_length):
-                next_hidden, (h, c) = self.lstm(image_hidden, (h, c))
-                unrolled_hidden.append(next_hidden)
-                image_hidden = next_hidden
-            image_hidden = torch.stack(unrolled_hidden, dim=1).squeeze()
-        elif self.unroll == 'transformer':
-            image_hidden = self.transformer(image_hidden)
+        # This is the input into gumbel generation; we will either unroll the
+        # hidden image (either via recurrent unrolling or transformer unrolling)
+        # or unroll text
+        #
+        # If decoder_input_ids is not None then we are performing caption
+        # training and will skip the gumbel generation
+        crossattention_input = None
+        batch_size = None
+
+        # Choose text vs caption-training vs image in the following order. Based
+        # on config values, multiple conditions may be true during training. We
+        # opt to enforce a strict ordering below to decide if EC should be
+        # run on text then caption-training then images.
+        text_unrolling_condition = sender_input_text is not None
+        caption_training_condition = decoder_input_ids is not None
+        image_unrolling_condition = sender_image is not None
+
+        if text_unrolling_condition:
+            batch_size = sender_input_text.size(0)
+            sender_input_encodings = self.encoder(
+                input_ids=sender_input_text,
+                attention_mask=sender_attention_mask
+            ).last_hidden_state
+            crossattention_input = sender_input_encodings
+        elif caption_training_condition:
+            sender_input_encodings = self.encoder(
+                input_ids=decoder_input_ids,
+                attention_mask=kwargs['caption_mask']
+            ).last_hidden_state
+            crossattention_input = sender_input_encodings
+        elif image_unrolling_condition:
+            batch_size = sender_image.size(0)
+            if len(sender_image.shape) == 2:
+                sender_image = sender_image.unsqueeze(1)
+            if self.unroll == 'recurrent':
+                unrolled_hidden = []
+                h = torch.zeros_like(sender_image).transpose(0, 1).to(
+                    sender_image.device
+                )
+                c = torch.zeros_like(sender_image).transpose(0, 1).to(
+                    sender_image.device
+                )
+                for i in range(self.unroll_length):
+                    next_hidden, (h, c) = self.lstm(sender_image, (h, c))
+                    unrolled_hidden.append(next_hidden)
+                    sender_image = next_hidden
+                sender_image = torch.stack(unrolled_hidden, dim=1).squeeze()
+            elif self.unroll == 'transformer':
+                sender_image = self.transformer(sender_image)
+            crossattention_input = sender_image
```

Comment:
Great work here!  There's probably a uniform API here for an `unroller`, and `self.unroller`, where the LSTM vs transformer logic happens elsewhere, but I just want to flag that as something to think about for the future; doesn't need to be added to this PR

Reply:
Makes sense! There's a lot of config overlap that makes this messy today but a good candidate to throw some inheritance into the mix in the future :-)